yeat/Snakefile

```diff
@@ -100,9 +100,11 @@ rule spades:
         read1="seq/downsample/{sample}.R1.fq.gz",
         read2="seq/downsample/{sample}.R2.fq.gz"
     threads: 128
+    params:
+        extra_args=config["extra_args"]
     shell:
         """
-        spades.py -1 {input.read1} -2 {input.read2} -t {threads} -o analysis/spades
+        spades.py -1 {input.read1} -2 {input.read2} -t {threads} -o analysis/spades {params.extra_args[spades]}
```

Comment:
Tried the:

>params:
         extra_args=config["extra_args"]["spades"]

The code works; however, after messing with the code for a while, I came to learn that all parameters in snakemake rules are initialized. Because of this, if the config file did not have a certain assembly algorithm, for example, `spades`, would fail on the `config["extra_args"]["spades"]` because the key does not exist. To prevent that, there is a few solutions that I could think of.

1) You can encapsulate the entire rule and only run if the key exists (but doing this option will open a whole new set of problems). Or,
2) Get the value from the dictionary only when the shell command is executed (meaning that the user's config has the desired assembly algorithm to run).

In the end, I went with the latter and learned that if `config["extra_args"]` does not have the key, snakemake will crash and fail. As a result, the following code access the key's value must be down in the shell command.

Reply:
Ah, good catch. So my suggested change doesn't make the code as concise as I thought. Thanks for the clear explanation. I think this is a good compromise.

Reply:
You more you know!